File: java/src/main/java/com/ngt/window/CountWindowAllReduceDemo.java
package com.ngt.window;

import org.apache.flink.api.common.functions.ReduceFunction;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.datastream.AllWindowedStream;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.windowing.windows.GlobalWindow;

/**
 * @author ngt
 * @create 2021-01-28 17:38
 * 不使用sum聚合，使用Reduce进行聚合
 */
public class CountWindowAllReduceDemo {
    public static void main(String[] args) throws Exception {
        Configuration conf = new Configuration();
        conf.setInteger("rest.port", 8181);
        StreamExecutionEnvironment env = StreamExecutionEnvironment.createLocalEnvironmentWithWebUI(conf);

        DataStreamSource<String> lines = env.socketTextStream("192.168.31.8", 8888);
        SingleOutputStreamOperator<Integer> nums = lines.map(Integer::parseInt);

        // 不分组，将整体当成一个组。并行度为1
        AllWindowedStream<Integer, GlobalWindow> window = nums.countWindowAll(4, 2);

        window.reduce(new ReduceFunction<Integer>() {
            @Override
            public Integer reduce(Integer value1, Integer value2) throws Exception {
                return value1 + value2;
            }
        }).print();
        env.execute();
    }
}

File: scala/src/main/scala/com/ngt/window/CountWindowAllReduceDemo.scala
package com.ngt.window

import org.apache.flink.api.common.functions.ReduceFunction
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.windowing.windows.GlobalWindow

/**
 * @author ngt
 * @create 2021-02-03 20:28
 * 不使用sum聚合，使用Reduce进行聚合
 */
object CountWindowAllReduceDemo {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    val lines: DataStream[String] = env.socketTextStream("192.168.31.8", 8888)

    val nums: DataStream[Int] = lines.map(_.toInt)
    val window: AllWindowedStream[Int, GlobalWindow] = nums.countWindowAll(4, 2)
    window.reduce(new ReduceFunction[Int] {
      override def reduce(value1: Int, value2: Int): Int = {
        value1 + value2
      }
    }).print()
    env.execute()
  }
}
